doc/proposals/leader-for-life.md

```diff
@@ -0,0 +1,58 @@
+# Leader for Life
+
+## Background
+
+Operators need leader election to ensure that if the same operator is running
+in two separate pods within the same namespace, only one of them will be
+active. The primary goal of leader election is to avoid contention between
+multiple active operators of the same type.
+
+High availability is not a goal of leader election for operators.
+
+Controller-runtime is adding leader election based on functionality present [in
+client-go](git@github.com:operator-framework/operator-sdk.git). However that
+implementation allows for the possibility of brief periods during which
+multiple leaders are active.
+
+Requirements have been [discussed on
+GitHub](https://github.com/operator-framework/operator-sdk/issues/136).
+
+This proposal is to add leader election to the SDK that follows a "leader for
+life" model, which does not allow for multiple concurrent leaders.
+
+## Goals
+
+* Provide leader election that is easy to use
+* Provide leader election that prohibits multiple leaders
```

Comment:
That would be characteristic of a lease-based approach. This leader-for-life approach avoids that complexity and any uncertainty about when or if a leader steps down.

Reply:
@xiang90 We're looking at a simplified leader for life approach.
In summary the leader pod creates a configmap(lock) with its ownerRef, and then when that leader pod exits/fails the k8s GC should remove that configmap(unlock) so that another pod could become the leader.
https://github.com/mhrivnak/leaderelection#simplified-algorithm

Let me know if you can think of any issues with this approach.

Reply:
@hasbro17 

From theoretical point of view, it is still the same as lease base approach and suffers from the same problem. 

It seems simple since, now we rely on the pod deletion, which MIGHT have causality relationship with the actual process death. The process can still be running even if the pod itself is deleted and the configmap is therefor GC-ed by k8s controller.

Reply:
My understanding is that after each process gets sent a TERM signal, it must either exit gracefully, or after a grace period it will be sent a KILL signal, before the Pod itself gets deleted. That's my interpretation of these docs:

https://kubernetes.io/docs/concepts/workloads/pods/pod/#termination-of-pods

So based on that, it seems there is no opportunity for the process to continue. Or is there some scenario I'm missing?

Reply:
This is what I mentioned to @hasbro17. It's possible the pod is deleted from the API forcefully, and the pod may still be running for an indeterminate amount of time if you don't wait for it to stop. I'm unsure if TERM is guaranteed to be sent if you force delete the pod.

Reply:
For force deletions(gracePeriod=0) the pod is deleted(removed from the API) before TERM or KILL is sent to the process. And even then there's an underlying grace period until the process is killed.
https://kubernetes.io/docs/concepts/workloads/pods/pod/#force-deletion-of-pods
`On the node, pods that are set to terminate immediately will still be given a small grace period before being force killed` 
So the process will still run after pod deletion for sure in that case.

In other cases, theoretically speaking, the kubelet will wait to KILL the process before deleting the Pod from the API according to:
https://kubernetes.io/docs/concepts/workloads/pods/pod/#termination-of-pods
But in practice, as xiang said, while pod deletion MIGHT have causality with process death, it's not guaranteed.

However based on the pod lifecycle, apart from force deletions, I'll have to trust that the process is killed before pod deletion. Unless we can identify a scenario in practice where this isn't true.

Reply:
What about the case of a pod being evicted due to a node not being able to reach the API server for an extended period of time? The pod should get stopped when able to reach the API again, though it's possible it's still acting.

I'm not 100% sure in the difference, but in the lease based election, your pod has a grace period and could potentially renew after getting connectivity.

In the for-life approach, I'm not sure, since I'm not 100% sure when the pod is actually deleted from the API during eviction.

Either way it's worth spending time researching and documenting more cases like this I think since there's a few that have already come up.

Reply:
@hasbro17 The node controller might just remove the node if inactive for long enough without confirming the pod is indeed stopped. There are quite a few cases I can think of. The correctness of this API will rely on the configuration on the Kubernetes cluster probably or the external environment. If you can control that config, it might not be a big issue.

Reply:
@chancez 

Exactly. You beat me by 1 minute :P.

Reply:
For force deletion, I agree that's not much of a concern. It's well-documented in multiple places that it is unsafe to use, and this case would be no exception.

Pulling a node's network connection is an interesting scenario to vet. This documentation seems to spell it out clearly that we do not need to be concerned:

https://kubernetes.io/docs/concepts/architecture/nodes/#condition

Regarding pods on an unreachable node: "the node controller does not force delete pods until it is confirmed that they have stopped running in the cluster. You can see the pods that might be running on an unreachable node as being in the Terminating or Unknown state. In cases where Kubernetes cannot deduce from the underlying infrastructure if a node has permanently left a cluster, the cluster administrator may need to delete the node object by hand. Deleting the node object from Kubernetes causes all the Pod objects running on the node to be deleted from the apiserver, and frees up their names."

Thank you all for thoroughly vetting this. :)

Reply:
@mhrivnak Right so if the node controller doesn't delete Pods from the apiserver for unreachable nodes then the configmap of the unreachable pod won't get GC'ed. And we won't have two leaders running at the same time.
We'll just be stuck with the partitioned leader pod until its node is reachable again, or an admin deletes the node.

We can keep documenting more cases where the pod-deletion=>process-stopped causality breaks down as we go along. But for now we can move forward with this approach.